lasy/propagators/collins_sfft_propagator.py

```diff
@@ -1,9 +1,283 @@
+        E_{i} (x,y,\omega) e^{ikS}dx_0dy_0,
+
+    where :math:`E_{i} (x,y,\omega)` is the initial/propagated fields complex field envelope
+    and :math:`S` is the propagator term
+
+    :math:
+        S = \bigg\{\frac{1}{2B}\Big[A(x_0^2+y_0^2)+D(x^2+y^2)-2(xx_0+yy_0)\Big]\bigg\},
+
+    defined in terms of the elements of the ``'ABCD'`` optical ray matrix.
+
+    Parameters
+    ----------
+    omega0 : float (in rad/s)
+        The center frequency of the laser field.
+
+    dim : string
+        Dimensionality of the array. Options are:
+
+        - ``'xyt'``: The laser pulse is represented on a 3D grid:
+                    Cartesian (x,y) transversely, and temporal (t) longitudinally.
+        - ``'rt'`` : The laser pulse is represented on a 2D grid:
+                    Cylindrical (r) transversely, and temporal (t) longitudinally.
+
+    abcd : 2d array
+        The 2D ray matrix of the optical system through which the beam propagates.
+        By default, this is initialised to be the unitary matrix:
+
+        .. math::
+
+            O =
+            \begin{pmatrix}
+            A & B \\
+            C & D
+            \end{pmatrix}=
+            \begin{pmatrix}
+            1 & 0 \\
+            0 & 1
+            \end{pmatrix}.
+    """
+
+    def __init__(self, omega0, dim, abcd=np.array([[1, 0], [0, 1]])):
         super().__init__()
-        print("empty init for CollinsSFFTPropagator")
+        self.update(dim=dim, omega0=omega0, abcd=abcd)
+
+    def update(self, dim, omega0, abcd):
+        r"""
+        Initialize or update the propagator if needed.
+
+        Parameters
+        ----------
+        dim : string
+            Dimensionality of the array. Options are:
+            - ``'xyt'``: Laser pulse represented on a 3D Cartesian grid.
+            - ``'rt'`` : Laser pulse represented on a 2D cylindrical grid.
+
+        omega0 : float (in rad.s^-1)
+            The main frequency :math:`\omega_0`, which is defined by the laser
+            wavelength :math:`\lambda_0`, as :math:`\omega_0 = 2\pi c/\lambda_0`.
+
+        abcd : 2d array
+            The 2D ray matrix of the optical system through which the beam propagates.
+            By default, this is initialised to be the unitary matrix:
+
+            .. math::
+
+                O =
+                \begin{pmatrix}
+                A & B \\
+                C & D
+                \end{pmatrix}=
+                \begin{pmatrix}
+                1 & 0 \\
+                0 & 1
+                \end{pmatrix}.
+        """
+        dim = dim if dim is not None else self.dim
+        assert isinstance(abcd, (np.ndarray))
+        assert dim in ["rt", "xyt"]
+
+        self.dim = dim
+        self.omega0 = omega0 if omega0 is not None else self.omega0
+        self.abcd = abcd  # optical ray matrix
+
+    def add_vacuum(self, distance):
+        vacuum = np.array([[1, distance], [0, 1]])
+        self.abcd = np.matmul(vacuum, self.abcd)
+        return self.abcd
```

Comment:
Probably no need to return it if the member variable is changed? That applies to several of these functions.